Parallel DeepSeg processes no longer try to use models which are in the process of being installed

SomeoneInParticular opened the pull request:
## Checklist

- [ ] **PR Sidebar**: I've filled in each of the options within the PR sidebar. ([Reviewers](https://github.com/spinalcordtoolbox/spinalcordtoolbox/wiki/Contributing:-Submitting-changes-to-SCT-(Opening-a-Pull-Request)#231-reviewers), [Assignees](https://github.com/spinalcordtoolbox/spinalcordtoolbox/wiki/Contributing:-Submitting-changes-to-SCT-(Opening-a-Pull-Request)#232-assignees), [Labels](https://github.com/spinalcordtoolbox/spinalcordtoolbox/wiki/Contributing:-Submitting-changes-to-SCT-(Opening-a-Pull-Request)#233-labels) (**exactly one** dark purple label!), [Milestone](https://github.com/spinalcordtoolbox/spinalcordtoolbox/wiki/Contributing:-Submitting-changes-to-SCT-(Opening-a-Pull-Request)#234-milestone), [Development](https://github.com/spinalcordtoolbox/spinalcordtoolbox/wiki/Contributing:-Submitting-changes-to-SCT-(Opening-a-Pull-Request)#235-development).)
- [ ] **Contributing Docs**: I've read the [Contributing guidelines](https://github.com/spinalcordtoolbox/spinalcordtoolbox/wiki/Contributing:-Submitting-changes-to-SCT-(Opening-a-Pull-Request)) to make sure my [branch](https://github.com/spinalcordtoolbox/spinalcordtoolbox/wiki/Contributing%3A-Submitting-changes-to-SCT-(Opening-a-Pull-Request)#part-1-branches) and [pull request](https://github.com/spinalcordtoolbox/spinalcordtoolbox/wiki/Contributing%3A-Submitting-changes-to-SCT-(Opening-a-Pull-Request)#part-2-pull-requests) meet SCT's guidelines. (Feel free to stop and fixup your commits before submitting.)
- [ ] **Tests**: I've added [relevant tests](https://github.com/spinalcordtoolbox/spinalcordtoolbox/wiki/Testing) for my contribution, if necessary. I've also made sure that my contribution passes [all of the automated tests](https://github.com/spinalcordtoolbox/spinalcordtoolbox/wiki/Contributing:-Submitting-changes-to-SCT-(Opening-a-Pull-Request)#24-checking-the-automated-tests) (which will be triggered after the PR is submitted).
- [ ] **Documentation**: I've updated the [relevant documentation](https://github.com/spinalcordtoolbox/spinalcordtoolbox/wiki/Documentation) for my changes, including argparse descriptions, docstrings, and ReadTheDocs tutorial pages.

## Description

As the title implies, via the use of a "mutex" semaphore, `sct_deepseg` processes run in parallel will no longer crash if/when they try to use a model that is in the process of being installed by another running process.

## Linked issues

Fixes #5084

SomeoneInParticular committed:
Refactored `sct_deepseg` main routine in preparation for install mutex integration (plus some general housekeeping); step 1.

SomeoneInParticular committed:
Removed unused import left in during experimentation.

SomeoneInParticular committed:
Corrected type hint formatting.

SomeoneInParticular committed:
Replaced `mutex` with `Mutex`, which retains most its properties while being much more flexible and easy to use.

SomeoneInParticular committed:
Fixed edge case when using a Mutex with a timeout shorter than its check interval

SomeoneInParticular committed:
Fixed formatting issues in documentation for "Mutex"

SomeoneInParticular committed:
Refactored `sct_deepseg` and `models` to utilize the new Mutex, improving their behaviour when run in parallel.

SomeoneInParticular committed:
Code quality fixes.

mguaypaq commented on spinalcordtoolbox/utils/fs.py:
```diff
@@ -264,22 +263,111 @@ def relpath_or_abspath(child_path, parent_path):
+        directory: str = tempfile.gettempdir(),
+        timeout: float | None = 5,  # 5 seconds
+        check_interval: float | None = 0.25,  # Every quarter second
+        fail_when_locked: bool | None = True,
```
Instead of hardcoding the defaults of (the current version of) portalocker's `BoundedSemaphore`, I would suggest using `**kwargs` and passing them on to `super().__init__` (if we want to consider `Mutex` to be a thin wrapper).

If we want to consider `Mutex` to be a self-contained abstraction instead, I would suggest hiding anything we don't actually use from `BoundedSemaphore` as an implementation detail.

(These suggestions are coming from a principle of trying to have a single source of truth, because duplicated truths tend to get out of sync. But, I can see an argument that it's very convenient to have local documentation, instead of having to dig around a dependency's source code for details.)

mguaypaq reviewed (commented):
Here are a few comments on the `Mutex` implementation, for now. I like your idea of logging a message when the lock isn't immediately available.

SomeoneInParticular committed:
Corrected erroneous timeout override

If the user provided a `timeout` of `0`, it would use the default timeout instead when it should not have.

Co-authored-by: Mathieu Guay-Paquet <mathieu.guay-paquet@polymtl.ca>

SomeoneInParticular commented on spinalcordtoolbox/utils/fs.py:
```diff
@@ -264,22 +263,111 @@ def relpath_or_abspath(child_path, parent_path):
+        """
+        # If we have a message on miss, do a very short acquisition attempt first
+        timeout = timeout if timeout else self.timeout
+        check_interval = check_interval if check_interval else self.check_interval
```
Thanks! Good catch.
